src/Program.ts

```diff
@@ -696,30 +753,36 @@ export class Program {
         };
     }
 
-    public async transpile(fileEntries: StandardizedFileEntry[], stagingFolderPath: string) {
-        let promises = Object.keys(this.files).map(async (filePath) => {
-            let file = this.files[filePath];
-            let filePathObj = fileEntries.find(x => s`${x.src}` === s`${file.pathAbsolute}`);
+    public async transpile(fileEntries: FileObj[], stagingFolderPath: string) {
+        const entries = Object.values(this.files).map(file => {
+            const filePathObj = fileEntries.find(x => s`${x.src}` === s`${file.pathAbsolute}`);
             if (!filePathObj) {
                 throw new Error(`Cannot find fileMap record in fileMaps for '${file.pathAbsolute}'`);
             }
             //replace the file extension
-            let outputCodePath = filePathObj.dest.replace(/\.bs$/gi, '.brs');
+            let outputPath = filePathObj.dest.replace(/\.bs$/gi, '.brs');
             //prepend the staging folder path
-            outputCodePath = s`${stagingFolderPath}/${outputCodePath}`;
-            let outputCodeMapPath = outputCodePath + '.map';
+            outputPath = s`${stagingFolderPath}/${outputPath}`;
+            return {
+                file: file,
+                outputPath: outputPath
+            };
+        });
 
-            let result = file.transpile();
+        this.plugins.emit('beforeTranspile', entries);
+
+        const promises = entries.map(async ({ file, outputPath }) => {
+            const result = file.transpile();
```

Comment:
There could be value in emitting a `beforeFileTranspile` and `afterFileTranspile` event in this function. It removes the need for plugin authors to walk every file themselves.

Reply:
done